src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -419,10 +423,38 @@ private static void BindProperty(PropertyInfo property, object instance, IConfig
 
             if (!type.IsValueType)
             {
-                bool hasDefaultConstructor = type.GetConstructors(DeclaredOnlyLookup).Any(ctor => ctor.IsPublic && ctor.GetParameters().Length == 0);
-                if (!hasDefaultConstructor)
+                ConstructorInfo[] constructors = type.GetConstructors(DeclaredOnlyLookup);
+
+                bool hasParameterlessPublicConstructor = constructors.Any(ctor => ctor.IsPublic && ctor.GetParameters().Length == 0);
```

Comment:
Sort by length, pick the last?

Reply:
That's what this is doing:

```csharp
  // find the biggest constructor so that we can bind to the most parameters
  ParameterInfo[] parameters = constructors[0].GetParameters();

  for (int index = 1; index < constructors.Length; index++)
  {
      ParameterInfo[] constructorParameters = constructors[index].GetParameters();
      if (constructorParameters.Length > parameters.Length)
      {
          parameters = constructorParameters;
      }
  }
```

I can look at some sort of `SortBy` method if needed, but I would guess that this is marginally better performing.

Reply:
Are constructors ordered.

Reply:
They don't appear to be. To make the test I added clearer, I created 4 overloads of the constructor in an arbitrary order, but ensuring that the smallest wasn't first. When calling `type.GetConstructors`, they appear to be in the order they were defined in source:

![image](https://user-images.githubusercontent.com/263416/160226310-84571adb-9e04-4bfd-a02c-5df6291dfd4d.png)

... which tallies up with:

![image](https://user-images.githubusercontent.com/263416/160226362-4e99b24b-01a7-417d-9fcb-f03eeb8a9fed.png)